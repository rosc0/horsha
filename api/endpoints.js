// Auth
export const TOKEN = '/token'
export const RESET_PASSWORD = '/password_reset/request'
export const REGISTER_DEVICE = '/device'

// Newsfeed
export const NEWSFEED = '/news'
export const JOURNAL_ENTRY = '/journal_entry'
export const COMMENTS = '/comment'
export const LIKE = '/like'

// User
export const USER = '/user'
export const USER_PROFILE = '/user/:id'
export const USER_SEARCH = '/user/search'
export const CURRENT_USER = '/user/me'
export const UPDATE_USER_PROFILE = '/user/:id/profile'
export const UPDATE_USER_PREFERENCES = '/user/:id/preferences'
export const UPDATE_USER_PROFILE_IMAGE = '/user/:id/profile_picture'
export const RESEND_EMAIL_CONFIRMATION =
  '/user/:userId/registered_email/:email/confirmation_state'
export const REGISTERED_EMAIL = '/user/:userId/registered_email'
export const SET_EMAIL_PRIMARY = '/user/:userId/primary_email'
export const CREATE_USER_WITH_FACEBOOK = '/user/facebook'

//Friends
export const REQUEST_FRIENDSHIP = '/user/:user_id/friend_request'
export const ACCEPT_FRIEND_REQUEST =
  '/user/:initiating_user_id/friend_request/:target_user_id/accept'
export const IGNORE_FRIEND_REQUEST =
  '/user/:initiating_user_id/friend_request/:target_user_id/state'
export const DELETE_FRIEND_REQUEST =
  '/user/:initiating_user_id/friend_request/:target_user_id'
export const DELETE_FRIEND = '/user/:user_id/friend/:target_user_id'
export const FRIEND_SUGGESTIONS = '/user/:user_id/friend_suggestions'

// Horses
export const HORSE = '/horse'
export const HORSE_USER = '/horse_user'
export const HORSE_PICTURE = '/profile_picture'
export const HORSE_ALBUM = '/horse_picture'
export const HORSE_USER_RELATION = '/relation'
export const HORSE_JOURNAL = '/journal_entry'
export const HORSE_LAST_ACCESSED = '/horse_user/:id/last_accessed'
export const STATUS_UPDATE = '/status_update'

// Rides
export const RIDE = '/ride'
export const GEO = '/geo/track/gpx'
export const TRACK = '/geo/gpx/track'
export const CONVERT_GPX_WAYLIST = '/geo/gpx/waypoint_list'
// Trails
export const TRAILS = '/trail'
export const TRAIL_PICTURES = '/trail_picture'
export const TRAIL_REVIEW = '/trail_review'
export const TRAIL_POI = '/trail_poi'
export const TRAIL_FAVORITES = '/trail_favorite'
export const TRAIL_RIDES = '/trail_ride'

// Point of Interests
export const POIS = '/poi'

// Report
export const REPORT = '/report'

// Notifications
export const READ_MARKER = '/notification/read_state'
export const NOTIFICATIONS = '/notification'

// Upload
export const UPLOAD = '/upload'
