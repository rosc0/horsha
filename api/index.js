import config from 'react-native-config'
import { build, serialize, emptyToNull } from '@application/utils'
import store from '@application/store'
import TimerMixin from 'react-timer-mixin'
import { setAlertStatus } from '../actions/app'
// import { logoutCurrentUser } from '../actions'
import * as k from '@constants/storageKeys'
import { AsyncStorage } from 'react-native'
// import Authentication from '@api/auth'

const title = 'There’s been a server error'
const subTitle = 'Something’s wrong, please try again later'

// const Auth = new Authentication()

class API {
  constructor(props) {
    this._config = new WeakMap()
    this._endpoint = new WeakMap()
    this._hasBaseURL = false

    if (!props || !props.isAuth) this._setup()
  }

  _setup() {
    const headers = {
      'Content-Type': 'application/json',
    }

    this._config.set(this, { headers })
  }

  setAuthConfig() {
    const headers = {
      'Content-Type': 'application/x-www-form-urlencoded',
    }

    this._config.set(this, { headers })
    return this
  }

  setContentType(contentType) {
    if (contentType) {
      const headers = {
        'Content-Type': contentType,
      }

      this._config.set(this, { headers })
    }
    return this
  }

  _setMethod(method) {
    const config = this._config.get(this) || {}
    config.method = method

    this._config.set(this, config)
  }

  setEndpoint(endpoint, hasBaseURL = false) {
    this._endpoint.set(this, endpoint)

    this._hasBaseURL = hasBaseURL
    return this
  }

  getEndpointPath() {
    return this._endpoint.get(this)
  }

  getEndpoint() {
    if (this._hasBaseURL) {
      return this._endpoint.get(this)
    }

    return config.API_BASE + this._endpoint.get(this)
  }

  getHeaders = () => {
    const state = store.getState()

    const {
      auth: { auth },
      user: { user },
    } = state

    return {
      access_token: auth.accessToken,
      user_id: user.id,
    }
  }

  setBody(body, isJSON) {
    const config = this._config.get(this) || {}

    config.body = isJSON
      ? JSON.stringify(emptyToNull(body))
      : serialize(emptyToNull(body))

    this._config.set(this, config)

    return this
  }

  getConfig() {
    let _config = this._config.get(this)

    const isAuth =
      _config.headers['Content-Type'] === 'application/x-www-form-urlencoded'

    if (isAuth) {
      let isClientIdDuplicated = false

      if (_config.body && !_config.body._parts) {
        isClientIdDuplicated = _config.body.indexOf('client_id') > -1
      }

      if (
        _config.method !== 'GET' &&
        isAuth &&
        !isClientIdDuplicated &&
        !_config.body._parts
      ) {
        _config.body += '&client_id=' + config.CLIENT_ID
      }
    }

    return this._config.get(this)
  }

  _makeRequest() {
    return new Promise((resolve, reject) => {
      let config = this.getConfig()
      let endpoint = this.getEndpoint()

      if (config.method === 'GET') {
        delete config.body
      }

      fetch(endpoint, config)
        .then(response => {
          TimerMixin.setTimeout(() => null, 0)

          if (response.status === 500) {
            // to handler if API return is not valid
            console.warn('throw Error', response)
            store.dispatch(setAlertStatus('error', title, subTitle))
            throw Error(response.statusText)
          }

          if (response.status === 403) {
            console.warn('logged out', response)
            if (
              response._bodyInit !==
              '{"status":403,"description":"User not allowed to allowed to see pictures for horse"}'
            ) {
              AsyncStorage.removeItem(k.AUTH_NEW_TOKENS)

              return store.dispatch({
                type: 'RESET',
                payload: {
                  app: store.getState().app,
                },
              })
            }
            // return response
            AsyncStorage.removeItem(k.AUTH_NEW_TOKENS)

            return store.dispatch({
              type: 'RESET',
              payload: {
                app: store.getState().app,
              },
            })
          }

          if (response.status === 401) {
            if (
              response._bodyInit === '{"code":401,"error":"Invalid token"} '
            ) {
              // ugly workaround
              AsyncStorage.removeItem(k.AUTH_NEW_TOKENS)
              return store.dispatch({
                type: 'RESET',
                payload: {
                  app: store.getState().app,
                },
              })
            }

            return response
          }

          if (response.status === 422) {
            if (
              response._bodyInit === '{"code":422,"text":"Missing user_id"}' ||
              response._bodyInit ===
                '{"status":422,"description":"Invalid UUID string: null"}'
            ) {
              // ugly workaround
              AsyncStorage.removeItem(k.AUTH_NEW_TOKENS)
              return store.dispatch({
                type: 'RESET',
                payload: {
                  app: store.getState().app,
                },
              })
            }

            return response
          }

          if (response.status < 200 || response.status > 299) {
            // to handler if API return is not valid
            console.warn('throw Error', response)
            store.dispatch(setAlertStatus('error', title, subTitle))
            throw Error('Error')
          }

          if (response.status === 204) {
            return response
          }

          const isResponseJSON =
            response.headers.get('content-type') &&
            response.headers.get('content-type').indexOf('application/json') !==
              -1

          return (response.status === 200 || response.status === 201) &&
            isResponseJSON
            ? response.json()
            : response
        })
        .then(res => {
          resolve(res)
        })
        .catch(err => {
          reject(err)
        })

      this._setup()
    })
  }

  addToken(type) {
    const { auth } = store.getState().auth
    const { accessToken: token, expireDate: expires } = auth

    const method =
      type === 'qs'
        ? this.addQueryStringToken.bind(this)
        : this.addJSONToken.bind(this)

    if (token && expires > +new Date()) {
      method(token)
    }
  }

  addQueryStringToken(token) {
    const body = this.getEndpointPath()
    const endpoint = build('get', { body, token })

    this.setEndpoint(endpoint)
  }

  addJSONToken(token) {
    const { body } = this.getConfig()
    const data = build('post', { body, token })

    this._config.set(this, { body: data })
  }

  get(hasToken) {
    if (hasToken) this.addToken('qs')

    this._setMethod('GET')
    this.setBody(null)
    return this._makeRequest()
  }

  post(hasToken) {
    if (hasToken) this.addToken('json')

    this._setMethod('POST')
    return this._makeRequest()
  }

  put(hasToken) {
    if (hasToken) this.addToken('json')

    this._setMethod('PUT')
    return this._makeRequest()
  }

  delete(hasToken, type = 'qs') {
    if (hasToken) this.addToken(type)

    this._setMethod('DELETE')
    return this._makeRequest()
  }
}

export default API
