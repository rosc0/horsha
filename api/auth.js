import { AsyncStorage } from 'react-native'
import * as k from '@constants/storageKeys'
import API from './index'
import * as endpoints from './endpoints'
import { USER_CHANGE_PASSWORD } from '../apollo/mutations/UserCollection'
import { client } from '../apollo/init'

class Authentication extends API {
  constructor({ isAuth = true } = {}) {
    super({ isAuth })
  }

  authenticate(credentials) {
    return new Promise((resolve, reject) => {
      this.setAuthConfig()
        .setEndpoint(endpoints.TOKEN)
        .setBody(credentials)
        .post()
        .then(response =>
          response.access_token ? this.storeToken(response) : response
        )
        .then(resolve)
        .catch(error => {
          this.handleError(error)
          reject()
        })
    })
  }

  refreshToken(refresh_token) {
    const body = {
      grant_type: 'refresh_token',
      refresh_token,
    }

    return new Promise((resolve, reject) => {
      this.setAuthConfig()
        .setEndpoint(endpoints.TOKEN)
        .setBody(body)
        .post()
        .then(this.storeToken)
        .then(resolve)
        .catch(error => {
          this.handleError(error)
          reject()
        })
    })
  }

  async isUserAuthenticated() {
    const authTokenData = await this.getToken()
    if (!authTokenData || !authTokenData.accessToken) {
      console.warn('no token')
      // throw new Error
      this.handleError('no token')
      return false
    }

    if (authTokenData.expireDate > +new Date()) {
      return authTokenData
    }

    return this.refreshToken(authTokenData.refreshToken)
  }

  async getToken() {
    const token = await AsyncStorage.getItem(k.AUTH_NEW_TOKENS)

    if (token) {
      return JSON.parse(token)
    }

    return false
  }

  storeToken(response) {
    const payload = {
      accessToken: response.access_token,
      expireDate: +new Date() + parseInt(response.expires_in) * 1000,
      refreshToken: response.refresh_token,
    }

    try {
      AsyncStorage.setItem(k.AUTH_NEW_TOKENS, JSON.stringify(payload))
    } catch (e) {
      this.handleError(e)
    }

    return payload
  }

  clearStorage() {
    return AsyncStorage.clear()
  }

  async resetPassword(password) {
    const { user_id } = this.getHeaders()

    const a = await client.mutate({
      mutation: USER_CHANGE_PASSWORD,
      variables: {
        password,
        userId: user_id,
      },
    })

    return a.userChangePassword
  }

  handleError(error) {
    console.warn('[AUTH] ERROR. ', error)
  }
}

export default Authentication
