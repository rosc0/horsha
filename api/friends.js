import API from './index'
import * as endpoints from './endpoints'
import { serialize } from '../utils'
import { client } from '../apollo/init'
import {
  GET_PEOPLE_YOU_MAY_KNOW,
  GET_PENDING_INVITATIONS,
  GET_IGNORED_INVITATIONS,
  GET_FRIENDS,
} from '../apollo/queries/UserCollection'
import {
  USER_FRIEND_REQUEST,
  USER_UNFRIEND_REQUEST,
  DENY_FRIEND_REQUEST,
  ACCEPT_FRIEND_REQUEST,
  UPDATE_FRIEND_REQUEST,
  IGNORE_FRIEND_REQUEST,
} from '../apollo/mutations/UserCollection'

class Friends extends API {
  async getFriends(userId = null, cursor = null, order_by = null) {
    const { user_id } = this.getHeaders()

    const {
      data: { userList },
    } = await client.query({
      query: GET_FRIENDS,
      variables: {
        userId: user_id,
        maxItems: 50,
      },
    })

    return userList
  }

  searchFriends(searchTerm, cursor) {
    const { access_token, user_id } = this.getHeaders()

    let params = {
      access_token,
      friends_with_user_id: user_id,
      order_by: '+name',
    }

    if (searchTerm) {
      params.name = searchTerm + '*'
    }

    if (cursor) {
      params.cursor = cursor
    }

    const endpoint = `${endpoints.USER}?${serialize(params)}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get()
        .then(resolve)
        .catch(reject)
    })
  }

  async getInvitations(state = 'pending') {
    const { user_id } = this.getHeaders()

    const {
      data: { userList },
    } = await client.query({
      query:
        state === 'pending' ? GET_PENDING_INVITATIONS : GET_IGNORED_INVITATIONS,
      variables: {
        userId: user_id,
        maxItems: 15,
      },
    })

    return userList
  }

  async getSuggestions() {
    const { user_id } = this.getHeaders()

    const {
      data: { userFriendSuggestions },
    } = await client.query({
      query: GET_PEOPLE_YOU_MAY_KNOW,
      variables: {
        userId: user_id,
        maxItems: 15,
      },
    })

    return userFriendSuggestions
  }

  async acceptRequest(initiatorUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: ACCEPT_FRIEND_REQUEST,
      variables: {
        initiatorUserId,
        userId: user_id,
      },
    })
  }

  async ignoreRequest(initiatorUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: IGNORE_FRIEND_REQUEST,
      variables: {
        initiatorUserId,
        userId: user_id,
      },
    })
  }

  async removeIgnoreRequest(initiatorUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: UPDATE_FRIEND_REQUEST,
      variables: {
        initiatorUserId,
        userId: user_id,
      },
    })
  }

  async deleteRequest(initiatorUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: DENY_FRIEND_REQUEST,
      variables: {
        initiatorUserId,
        userId: user_id,
      },
    })
  }

  async addFriend(targetUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: USER_FRIEND_REQUEST,
      variables: {
        targetUserId,
        userId: user_id,
      },
    })
  }

  async removeFriend(targetUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: USER_UNFRIEND_REQUEST,
      variables: {
        targetUserId,
        userId: user_id,
      },
    })
  }

  async followUser(targetUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: USER_FOLLOW_REQUEST,
      variables: {
        targetUserId,
        userId: user_id,
      },
    })
  }

  async unfollowUser(targetUserId) {
    const { user_id } = this.getHeaders()

    client.mutate({
      mutation: USER_UNFOLLOW_REQUEST,
      variables: {
        targetUserId,
        userId: user_id,
      },
    })
  }

  handleError(err) {
    // console.log('[Friends] ERROR: ', err)
  }
}

export default Friends
