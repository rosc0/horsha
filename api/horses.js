import API from './index'
import {
  serialize,
  formatBirthday,
  formatHeightUnit,
  formatWeightUnit,
} from '@application/utils'
import * as endpoints from './endpoints'
import store from '@application/store'
import {
  POST_PUBLISH,
  POST_ATTACH_MEDIA,
} from '../apollo/mutations/PostCollection'
import { client } from '../apollo/init'
import {
  GET_NEWSFEED_ITEM,
  GET_NEWSFEED_COLLECTION,
} from '../apollo/queries/NewsFeedCollection'

class Horses extends API {
  getHorses(limit = 50, cursor = false) {
    let params = {
      ...this.getHeaders(),
      max_items: limit,
    }

    if (cursor) params.cursor = cursor

    const endpoint = `${endpoints.HORSE_USER}?${serialize(params)}`

    return this.setEndpoint(endpoint).get()
  }

  getHorsesByUser(user_id) {
    const { access_token } = this.getHeaders()

    const params = serialize({ access_token, user_id })

    const endpoint = `${endpoints.HORSE_USER}?${params}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get()
        .then(resolve)
        .catch(reject)
    })
  }

  addHorse(horseData) {
    const { access_token, user_id: owner_user_id } = this.getHeaders()
    const state = store.getState()
    const { unit_system, height_unit } = state.user.user.account.preferences

    const {
      name,
      birthday,
      gender,
      color,
      breed,
      weight,
      height,
      description = null,
      horseCare: care_info,
    } = horseData

    const params = serialize({
      access_token,
    })

    const body = {
      name,
      birthday: formatBirthday(birthday),
      gender,
      color,
      breed,
      weight: weight
        ? {
            unit: formatWeightUnit(unit_system),
            value: Number(weight),
          }
        : null,
      height: height
        ? {
            unit: formatHeightUnit(height_unit),
            value: Number(height),
          }
        : null,
      description,
      care_info,
      owner_user_id,
    }
    const endpoint = `${endpoints.HORSE}?${params}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .setBody({ profile: { ...body } }, true)
        .post()
        .then(resolve)
        .catch(reject)
    })
  }

  editHorseUser({ horseUserId, ...data }) {
    const { access_token } = this.getHeaders()

    const params = serialize({
      access_token,
    })

    const endpoint = `${endpoints.HORSE_USER}/${horseUserId}${
      endpoints.HORSE_USER_RELATION
    }?${params}`
    return this.setEndpoint(endpoint)
      .setBody(data, true)
      .put()
  }

  addHorseImage(horse_id, upload_key) {
    const { access_token, user_id } = this.getHeaders()

    const params = serialize({
      access_token,
    })

    const endpoint = `${endpoints.HORSE_ALBUM}?${params}`

    const body = {
      upload_key,
      user_id,
      horse_id,
    }

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .setBody(body, true)
        .post()
        .then(resolve)
        .catch(reject)
    })
  }

  async deleteHorseImage(imagesId) {
    const { access_token } = this.getHeaders()

    const params = serialize({
      access_token,
    })

    return Promise.all(
      imagesId.map(imageId =>
        this.setEndpoint(
          `${endpoints.HORSE_ALBUM}/${imageId}?${params}`
        ).delete()
      )
    )
  }

  async addPostImage(postId, media) {
    try {
      const saved = await client.mutate({
        mutation: POST_ATTACH_MEDIA,
        variables: { postId, media },
      })

      return saved.data.postAttachMedia
    } catch (error) {
      console.error(error)
      return error
    }
  }

  setHorseProfileImage(horseId, upload_key) {
    const { access_token } = this.getHeaders()

    const params = serialize({
      access_token,
    })

    const endpoint = `${endpoints.HORSE}/${horseId}${
      endpoints.HORSE_PICTURE
    }?${params}`

    const body = {
      upload_key,
    }

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .setBody(body, true)
        .put()
        .then(resolve)
        .catch(reject)
    })
  }

  getHorseProfile(horseId) {
    const headers = this.getHeaders()
    const params = {
      ...headers,
      horse_id: horseId,
    }

    const endpoint = `${endpoints.HORSE}/${horseId}?${serialize(params)}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get()
        .then(resolve)
        .catch(reject)
    })
  }

  editHorse(id, content) {
    const { access_token } = this.getHeaders()
    const state = store.getState()
    const { unit_system, height_unit } = state.user.user.account.preferences

    const params = serialize({ access_token })

    const endpoint = `${endpoints.HORSE}/${id}/profile?${params}`

    const body = {
      id,
      name: content.name,
      breed: content.breed,
      color: content.color,
      gender: content.gender,
      weight: content.weight
        ? {
            unit: formatWeightUnit(unit_system),
            value: Number(content.weight),
          }
        : null,
      height: content.height
        ? {
            unit: formatHeightUnit(height_unit),
            value: Number(content.height),
          }
        : null,
      birthday: formatBirthday(content.birthday),
      description: content.description,
      care_info: content.horseCare && content.horseCare,
    }

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .setBody(body, true)
        .put()
        .then(resolve)
        .catch(reject)
    })
  }

  getHorseUser(horseId, userId = null) {
    const { access_token } = this.getHeaders()
    const params = {
      access_token,
      horse_id: horseId,
    }

    if (userId) {
      params.user_id = userId
    }

    const endpoint = `${endpoints.HORSE_USER}?${serialize(params)}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get()
        .then(resolve)
        .catch(reject)
    })
  }

  getHorseAlbum(horseId, filter = 'all_photos') {
    const { access_token, user_id } = this.getHeaders()

    let params = {
      horse_id: horseId,
      access_token,
    }

    if (filter === 'only_mine') {
      params.user_id = user_id
    }

    const endpoint = `${endpoints.HORSE_ALBUM}?${serialize(params)}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get()
        .then(resolve)
        .catch(reject)
    })
  }

  getHorseJournal(horseId) {
    const { access_token } = this.getHeaders()
    const data = serialize({ horse_id: horseId, max_items: 50, access_token })
    const endpoint = `${endpoints.HORSE_JOURNAL}?${data}`
    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get()
        .then(resolve)
        .catch(reject)
    })
  }

  async saveHorseJournal({ postId, horseId, shareScope, text, media }) {
    const { user_id } = this.getHeaders()

    const saved = await client.mutate({
      mutation: POST_PUBLISH,
      variables: { postId, horseId, shareScope, text, media },

      refetchQueries: () => [
        {
          query: GET_NEWSFEED_COLLECTION,
          variables: {
            userId: user_id,
            maxItems: 10,
          },
        },
        {
          query: GET_NEWSFEED_ITEM,
          variables: {
            postId,
          },
        },
      ],
    })
    return saved.data.postPublishContent
  }

  async editHorseJournal({ postId, horseId, shareScope, text, media }) {
    const { user_id } = this.getHeaders()

    const saved = await client.mutate({
      mutation: POST_PUBLISH,
      variables: { postId, horseId, shareScope, text, media },
      refetchQueries: () => [
        {
          query: GET_NEWSFEED_ITEM,
          variables: {
            postId,
          },
        },
      ],
    })
    return saved.data.postPublishContent
  }

  editShareScope(id, shareScope) {
    const { access_token } = this.getHeaders()
    const params = serialize({ access_token })

    const endpoint = `${endpoints.HORSE_JOURNAL}/${id}/share_scope?${params}`

    const payload = {
      share_scope: shareScope,
    }

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .setBody(payload, true)
        .put()
        .then(resolve)
        .catch(reject)
    })
  }

  addFriendTeam(horseId, userId) {
    const { access_token } = this.getHeaders()

    const body = {
      horse_id: horseId,
      user_id: userId,
    }

    const params = serialize({
      access_token,
    })

    const endpoint = `${endpoints.HORSE_USER}?${params}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .setBody(body, true)
        .post()
        .then(resolve)
        .catch(reject)
    })
  }

  getHorseUserRelation(horse_id) {
    const { access_token, user_id } = this.getHeaders()

    const params = serialize({ access_token, user_id, horse_id })

    const endpoint = `${endpoints.HORSE_USER}?${params}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get()
        .then(resolve)
        .catch(reject)
    })
  }

  setLatestHorse(horseUserId) {
    const { access_token } = this.getHeaders()

    const params = serialize({
      access_token,
    })

    const endpoint = `${endpoints.HORSE_LAST_ACCESSED.replace(
      ':id',
      horseUserId
    )}?${params}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .setBody(null, true)
        .put()
        .then(resolve)
        .catch(reject)
    })
  }

  deleteHorseUser(horseId) {
    const { access_token } = this.getHeaders()

    const params = serialize({
      access_token,
    })

    const endpoint = `${endpoints.HORSE_USER}/${horseId}?${params}`

    return this.setEndpoint(endpoint).delete()
  }

  getJournalPost(journalId) {
    const { access_token } = this.getHeaders()

    const params = serialize({
      access_token,
    })

    const endpoint = `${endpoints.HORSE_JOURNAL}/${journalId}?${params}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get()
        .then(resolve)
        .catch(reject)
    })
  }

  getStatusPost(statusId) {
    const { access_token, user_id } = this.getHeaders()

    const params = serialize({
      access_token,
      relation_with_user_id: user_id,
    })

    const endpoint = `${endpoints.STATUS_UPDATE}/${statusId}?${params}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get()
        .then(resolve)
        .catch(reject)
    })
  }
}

export default Horses
