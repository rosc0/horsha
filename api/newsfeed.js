import API from './index'
import { serialize } from '@application/utils'
import * as endpoints from './endpoints'
import config from 'react-native-config'
import {
  POST_LIKE,
  POST_ADD_COMMENT,
  POST_DELETE_COMMENT,
  POST_UNLIKE,
  POST_EDIT_COMMENT,
} from '../apollo/mutations/PostCollection'
import { client } from '../apollo/init'
import { GET_POST_COMMENTS } from '../apollo/queries/PostCollection'
import { GET_NEWSFEED_ITEM } from '../apollo/queries/NewsFeedCollection'

class Newsfeed extends API {
  getUpdates(maxItems = 10, cursor = false) {
    const { user_id } = this.getHeaders()
    let data = { user_id, max_items: maxItems }

    if (cursor) {
      data['cursor'] = cursor
    }

    const endpoint = `${endpoints.NEWSFEED}?${serialize(data)}`
    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get({ token: true })
        .then(resolve)
        .catch(reject)
    })
  }

  getUpdate(id) {
    const newsId = id ? `/${id}` : ''
    const endpoint = `${endpoints.JOURNAL_ENTRY}${newsId}`

    return this.setEndpoint(endpoint).get({ token: true })
  }

  deleteUpdate(id) {
    const { access_token } = this.getHeaders()

    const params = serialize({ access_token })
    const endpoint = `${endpoints.JOURNAL_ENTRY}/${id}?${params}`

    return this.setEndpoint(endpoint).delete()
  }

  async toggleLike(likeState, postId) {
    const { user_id } = this.getHeaders()

    const data = await client.mutate({
      mutation: likeState ? POST_UNLIKE : POST_LIKE,
      variables: {
        postId,
        userId: user_id,
      },
      refetchQueries: () => [
        {
          query: GET_NEWSFEED_ITEM,
          variables: {
            postId,
          },
        },
      ],
    })
    return data
  }

  getComments(
    contentId,
    contentType,
    previous = null,
    next = null,
    commentId = null,
    maxItems = null
  ) {
    // /comment?
    // max_items=2
    // &journal_entry_id=d1798c15-e9b4-4b64-abee-452c8e138a85
    // &cursor={next-value}

    let data = {
      max_items: maxItems ? maxItems : 5,
    }

    if (contentType === 'journal_entry') {
      data.journal_entry_id = contentId
    }

    if (contentType === 'status_update') {
      data.status_update_id = contentId
    }

    if (commentId) {
      data.from_id = commentId
    }

    if (next) {
      data.cursor = next
      data.page_dir = 'b'
    }

    if (previous) {
      data.cursor = previous
    }

    const endpoint = `${endpoints.COMMENTS}?${serialize(data)}`

    return new Promise((resolve, reject) => {
      this.setEndpoint(endpoint)
        .get({ token: true })
        .then(resolve)
        .catch(reject)
    })
  }

  async addComment(postId, text) {
    const { userId } = this.getHeaders()

    const data = await client.mutate({
      mutation: POST_ADD_COMMENT,
      variables: {
        postId,
        text,
        userId,
      },
      refetchQueries: () => [
        {
          query: GET_NEWSFEED_ITEM,
          variables: {
            postId,
          },
        },
        {
          query: GET_POST_COMMENTS,
          variables: {
            postId,
          },
        },
      ],
    })
    return data
  }

  async updateComment(commentId, postId, text) {
    const { userId } = this.getHeaders()

    const data = await client.mutate({
      mutation: POST_EDIT_COMMENT,
      variables: {
        postId,
        commentId,
        text,
      },
      refetchQueries: () => [
        {
          query: GET_NEWSFEED_ITEM,
          variables: {
            postId,
          },
        },
        {
          query: GET_POST_COMMENTS,
          variables: {
            postId,
          },
        },
      ],
    })
    return data
  }

  async removeComment(commentId, postId) {
    const data = await client.mutate({
      mutation: POST_DELETE_COMMENT,
      variables: {
        postId,
        commentId,
      },
      refetchQueries: () => [
        {
          query: GET_NEWSFEED_ITEM,
          variables: {
            postId,
          },
        },
        {
          query: GET_POST_COMMENTS,
          variables: {
            postId,
          },
        },
      ],
    })
    console.log('delete', data)
    return data
  }
}

export default Newsfeed
