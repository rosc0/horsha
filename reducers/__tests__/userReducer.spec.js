import userReducer, { initialState as state } from '../user'
import types from '../../constants/actionTypes'

import { current_user } from '../../../test/api'

describe('User Reducer', () => {
  it('should return initial state', () => {
    expect(userReducer()).toBe(state)
    expect(userReducer()).toMatchSnapshot()
  })

  it('should get current user', () => {
    expect(
      userReducer(state, {
        type: types.CURRENT_USER_FULFILLED,
        payload: current_user.GET,
      })
    ).toMatchSnapshot()
  })
})
