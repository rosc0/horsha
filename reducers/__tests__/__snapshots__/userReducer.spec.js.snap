// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Reducer should get current user 1`] = `
Object {
  "error": false,
  "fetched": true,
  "fetching": false,
  "user": Object {
    "account_status": "enabled",
    "birthday": "1977-09-25",
    "country_code": "NL",
    "date_account_status_changed": 1519053985000,
    "date_created": 1519053985000,
    "date_verification_state_changed": 1519072823000,
    "friendship_requested": false,
    "has_password": true,
    "horse_membership_count": 32,
    "id": "CURRENT_USER_ID",
    "name": "Ronaldo Lima",
    "preferences": Object {
      "height_unit": "centimeters",
      "push_notification_settings": Object {
        "push_enabled": true,
        "topics": Object {
          "horse_team": true,
          "journal_entry": true,
          "trail_content": true,
        },
      },
      "unit_system": "metric",
    },
    "primary_email": "ronaldo@horsha.com",
    "profile_picture": Object {
      "height": 0,
      "url": "https://dev-image-transformation.horsha.com/193d4e08-ae70-4cde-9fdf-e8513fb836e1/transformation",
      "width": 0,
    },
    "profile_text": "Mdadwadwa↵Sea↵Wd↵A↵Sea↵Sea↵Sea↵D↵Wa↵Dwadwadwadwa↵Dwadwadwadwadwa↵Sea↵Dwadwadwadwadwa↵Kkkkjkjljljljljljjjjjjkjkldwadwadw↵Dwa↵D↵Aw↵Dwa↵Dwadwadwadwadwa↵Dwa↵D↵Wa↵Dwadwadwadwadwadwadwadwadwa",
    "registered_emails": Array [
      Object {
        "confirmed": false,
        "email": "dwa@dsa.com",
      },
      Object {
        "confirmed": false,
        "email": "hshdhdd@jsjsj.com",
      },
      Object {
        "confirmed": false,
        "email": "ronaldo@horsha.com",
      },
    ],
    "riding_totals": Object {
      "ascent": 0,
      "count": 137,
      "descent": 0,
      "distance": 1565929,
      "duration": 9075763417,
    },
    "role": "user",
    "verification_state": "verified",
  },
  "userSearch": Object {
    "fetching": false,
    "users": Array [],
  },
}
`;

exports[`User Reducer should return initial state 1`] = `
Object {
  "error": null,
  "fetched": false,
  "fetching": false,
  "user": Object {
    "account_status": null,
    "date_created": null,
    "gender": null,
    "has_password": null,
    "horse_membership_count": null,
    "id": null,
    "name": null,
    "nr_of_rides": null,
    "primary_email": null,
    "role": null,
    "total_riding_ascent": null,
    "total_riding_descent": null,
    "total_riding_distance": null,
    "total_riding_time": null,
    "verification_state": null,
  },
  "userSearch": Object {
    "fetching": false,
    "users": Array [],
  },
}
`;
